### PR TITLE
Gateway hpa based on cpu

### PR DESCRIPTION
**Pull request by tamirdavid1**

### CODE REVIEW

File: autoscaler/controllers/gateway/hpa.go [hunk 1]
```diff
@@ -35,9 +35,22 @@ func syncHPA(gateway *odigosv1.CollectorsGroup, ctx context.Context, c client.Cl
 
 	var hpa client.Object
 
-	memLimit := gateway.Spec.MemorySettings.GomemlimitMiB * memoryLimitPercentageForHPA / 100.0
+	// Memory metric calculation
+	memLimit := gateway.Spec.ResourcesSettings.GomemlimitMiB * memoryLimitPercentageForHPA / 100.0
 	metricQuantity := resource.MustParse(fmt.Sprintf("%dMi", memLimit))
 
+	// CPU metric calculation
+	cpuTargetMillicores := gateway.Spec.ResourcesSettings.CpuLimitMillicores * 75 / 100
```
Comment (blumamir) [inline autoscaler/controllers/gateway/hpa.go:43]:
```suggestion
	cpuTargetMillicores := gateway.Spec.ResourcesSettings.CpuLimitMillicores * 75 / 100.0
```

File: scheduler/controllers/clustercollectorsgroup/resource_config.go [hunk 2]
```diff
@@ -6,8 +6,19 @@ import (
...
 
+	// the default CPU request in millicores
+	defaultRequestCPUm = 500
+	// the default CPU limit in millicores
+	defaultLimitCPUm = 1000
+
+	// MinReplicasDefault is the default number of replicas for the collector
+	MinReplicasDefault = 1
+	// MaxReplicasDefault is the default maximum number of replicas for the collector hpa
+	MaxReplicasDefault = 10
```
Comment (blumamir) [inline scheduler/controllers/clustercollectorsgroup/resource_config.go:20]: Above we have `default***` should we use the same name convention here as well?
Also, should it start with Capital letter?

File: helm/odigos/values.yaml [hunk 3]
```diff
@@ -25,6 +25,25 @@ collectorGateway:
...
+  # it will be embedded in the deployment as a resource limit
+  # of the form "cpu: <value>m".
+  # default value is 1000m
+  limitCPUm: 1000
+
+  # The number of replicas for the cluster gateway collector deployment.
+  # Also uses in MinReplicas the HPA config. 
+  MinReplicas: 1
+  # The maxReplicas in the HPA config.
+  MaxReplicas: 10
```
Comment (blumamir) [inline helm/odigos/values.yaml:44]:
```suggestion
  minReplicas: 1
  # The maxReplicas in the HPA config.
  maxReplicas: 10
```

File: scheduler/controllers/clustercollectorsgroup/resource_config.go [hunk 4]
```diff
@@ -27,37 +38,45 @@ const (
...
 		GomemlimitMiB:              gomemlimitMiB,
 	}
 }
+
+// Returns the value if it is greater than 0, otherwise returns the default value.
+func getOrDefault[T ~int | ~int32 | ~int64 | ~float32 | ~float64](config *common.CollectorGatewayConfiguration, value T, defaultValue T) T {
+	if config != nil && value > 0 {
+		return value
+	}
+	return defaultValue
```
Comment (blumamir) [inline scheduler/controllers/clustercollectorsgroup/resource_config.go:81]: Why do we need the `config` here? looks like it's unused

File: scheduler/controllers/clustercollectorsgroup/resource_config.go [hunk 5]
```diff
@@ -27,37 +38,45 @@ const (
...
-	memoryRequestMiB := defaultRequestMemoryMiB
-	if odigosConfig.CollectorGateway != nil && odigosConfig.CollectorGateway.RequestMemoryMiB > 0 {
-		memoryRequestMiB = odigosConfig.CollectorGateway.RequestMemoryMiB
-	}
-
-	memoryLimitMiB := int(float64(memoryRequestMiB) * memoryLimitAboveRequestFactor)
+func getGatewayResourceSettings(odigosConfig *common.OdigosConfiguration) *odigosv1.CollectorsGroupResourcesSettings {
+	gatewayConfig := odigosConfig.CollectorGateway
 
+	gatewayMinReplicas := getOrDefault(gatewayConfig, gatewayConfig.MinReplicas, MinReplicasDefault)
```
Comment (blumamir) [inline scheduler/controllers/clustercollectorsgroup/resource_config.go:46]: `gatewayConfig.MinReplicas` need `nil` check on `gatewayConfig` before dereferencing it

File: autoscaler/controllers/gateway/hpa.go (hunk 1 above)
Comment (tamirdavid1) [reply autoscaler/controllers/gateway/hpa.go:43]: why is needed?

File: autoscaler/controllers/gateway/hpa.go [hunk 6]
```diff
@@ -35,9 +36,26 @@ func syncHPA(gateway *odigosv1.CollectorsGroup, ctx context.Context, c client.Cl
...
+	memLimit := gateway.Spec.ResourcesSettings.GomemlimitMiB * memoryLimitPercentageForHPA / 100
 	metricQuantity := resource.MustParse(fmt.Sprintf("%dMi", memLimit))
 
+	// CPU metric calculation
+	cpuTargetMillicores := gateway.Spec.ResourcesSettings.CpuLimitMillicores * cpuLimitPercentageForHPA / 100
+	metricQuantityCPU := resource.MustParse(fmt.Sprintf("%dm", cpuTargetMillicores))
+
+	minReplicas := defaltMinReplicas
+	if gateway.Spec.ResourcesSettings.MinReplicas != nil &&
+		*gateway.Spec.ResourcesSettings.MinReplicas != int(*minReplicas) && *gateway.Spec.ResourcesSettings.MinReplicas > 0 {
```
Comment (blumamir) [inline autoscaler/controllers/gateway/hpa.go:49]:
```suggestion
	if gateway.Spec.ResourcesSettings.MinReplicas != nil && *gateway.Spec.ResourcesSettings.MinReplicas > 0 {
```

File: autoscaler/controllers/gateway/hpa.go [hunk 7]
```diff
@@ -35,9 +36,26 @@ func syncHPA(gateway *odigosv1.CollectorsGroup, ctx context.Context, c client.Cl
...
+		minReplicas = intPtr(int32(*gateway.Spec.ResourcesSettings.MinReplicas))
+	}
+
+	maxReplicas := defaultMaxReplicas
+	if gateway.Spec.ResourcesSettings.MaxReplicas != nil &&
+		*gateway.Spec.ResourcesSettings.MaxReplicas != int(maxReplicas) && *gateway.Spec.ResourcesSettings.MaxReplicas > 0 {
```
Comment (blumamir) [inline autoscaler/controllers/gateway/hpa.go:55]:
```suggestion
	maxReplicas := defaultMaxReplicas
	if gateway.Spec.ResourcesSettings.MaxReplicas != nil && *gateway.Spec.ResourcesSettings.MaxReplicas > 0 {
```

File: autoscaler/controllers/gateway/hpa.go [hunk 8]
```diff
@@ -22,11 +22,12 @@ import (
 
 const (
 	memoryLimitPercentageForHPA = 75
+	cpuLimitPercentageForHPA    = 75
 )
 
 var (
-	minReplicas                = intPtr(1)
-	maxReplicas                = int32(10)
+	defaltMinReplicas          = intPtr(1)
```
Comment (blumamir) [inline autoscaler/controllers/gateway/hpa.go:29]:
```suggestion
	defaultMinReplicas          = intPtr(1)
```

File: autoscaler/controllers/gateway/hpa.go [hunk 9]
```diff
@@ -22,11 +22,12 @@ import (
...
+	defaultMaxReplicas         = int32(10)
```
Comment (blumamir) [inline autoscaler/controllers/gateway/hpa.go:30]: these value describes the same thing (replicas) but one is `intPtr` and the other is `int32`. should they both be `int32`?

**Review by blumamir — APPROVED** [review]

**Pull request merged**